### DISCUSSION TITLE
use goroutines to accelerate encryption

### DISCUSSION THREAD

**Discussion by JohnJimAir** [tuneinsight/lattigo#433] [General]
`func Encrypt_Parallel(params hefloat.Parameters, encoder *hefloat.Encoder, encryptor *rlwe.Encryptor, values []float64, ) (values_ct []*rlwe.Ciphertext) {
	
	length := len(values)
	values_ct = make([]*rlwe.Ciphertext, length)
	
	var wg sync.WaitGroup
	for i:=0;i<length;i++ {
		wg.Add(1)
		go func(i int) {
			defer wg.Done()
			longvalues := make([]float64, params.MaxSlots())
			for j:=0;j<len(longvalues);j++ {
				longvalues[j] = values[i]
			}
			value_pt := hefloat.NewPlaintext(params, params.MaxLevel())
			if err := encoder.Encode(longvalues, value_pt); err != nil {
				panic(err)
			}
			value_ct, err := encryptor.EncryptNew(value_pt)
			if err != nil {
				panic(err)
			}
			values_ct[i] = value_ct
		}(i)
	}
	wg.Wait()
	return
}`
this func is meant to use goroutines to get multiple ciphertexts in parallel, but the result is odd. The ciphertextxs returned are some vary large numbers, where is wrong？

**Comment by ChristianMct** [comment]
The Lattigo API is not concurrency-safe. Consequently, you cannot use the same 'Encryptor' struct in multiple goroutines. 

Instead, you need to have a separate encryptor per goroutine. Note that you can obtain new encryptors efficiently by using the 'Encryptor.ShallowCopy()' method. 

Hope this helps

**Reply by JohnJimAir** [reply]
thank you vary much
